File: ga/4omini_result_2024_7/result/weakptr_l_mini/38_weakptr_l.cpp
use std::rc::{Rc, Weak};

struct MyStruct {
    value: i32,
}

impl MyStruct {
    fn new(v: i32) -> MyStruct {
        println!("Constructor");
        MyStruct { value: v }
    }
}

impl Drop for MyStruct {
    fn drop(&mut self) {
        println!("Destructor");
    }
}

fn observe(weak_ptr: &Weak<MyStruct>) {
    if let Some(shared_ptr) = weak_ptr.upgrade() {
        println!("Object is alive, value: {}", shared_ptr.value);
    } else {
        println!("Object has been destroyed");
    }
}

fn main() {
    let ptr1: Rc<MyStruct> = Rc::new(MyStruct::new(10));
    let weak_ptr1: Weak<MyStruct> = Rc::downgrade(&ptr1);

    {
        // Cloning the pointer to create a new reference without affecting the Weak pointer
        let ptr2 = Rc::clone(&ptr1);
        println!("Inside inner scope");
        observe(&weak_ptr1);

        let weak_ptr2: Weak<MyStruct> = Rc::downgrade(&ptr2);
        if let Some(shared_ptr) = weak_ptr2.upgrade() {
            shared_ptr.value = 20; // Modify the value
            println!("Value: {}", shared_ptr.value);
        }
    } // ptr2 goes out of scope here

    println!("Outside inner scope");
    observe(&weak_ptr1);

    // When we drop ptr1, it will also drop the MyStruct
    drop(ptr1);
    observe(&weak_ptr1);
}

File: ga/glm4_result_2024_6/result/weakptr_l_glm/38_weakptr_l.rs
use std::sync::{Arc, Weak};
use std::fmt;

#[derive(Clone)]
struct MyStruct {
    value: i32,
}

impl MyStruct {
    fn new(value: i32) -> Arc<Self> {
        Arc::new(MyStruct { value })
    }
}

impl fmt::Display for MyStruct {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Value: {}", self.value)
    }
}

fn observe(weak: &Weak<MyStruct>) {
    if let Some(strong) = weak.upgrade() {
        println!("{}", strong);
    } else {
        println!("Object has been destroyed");
    }
}

fn main() {
    let ptr1 = Arc::new(MyStruct::new(10));
    let weak_ptr1 = Arc::downgrade(&ptr1);

    {
        let ptr2 = ptr1.clone();
        println!("Inside inner scope");
        observe(&weak_ptr1);

        if let Some(strong) = weak_ptr1.upgrade() {
            strong.value = 20;
            println!("{}", strong);
        }
    }

    println!("Outside inner scope");
    observe(&weak_ptr1);

    ptr1 = Arc::new(MyStruct::new(30));
    observe(&weak_ptr1);
}